opendevin/runtime/od_runtime_client/runtime.py

```diff
@@ -0,0 +1,258 @@
+        # self.loop.run_until_complete(self._init_websocket_connect())
+    
+    async def _init_websocket_connect(self):
+        self.websocket = await websockets.connect(self.uri)
+    
+    def _init_sandbox(self):
+        try:
+            # start the container
+            mount_dir = config.workspace_mount_path
+            self.container = self.docker_client.containers.run(
+                self.container_image,
+                command='tail -f /dev/null',
+                # TODO: test it in mac and linux
+                # network_mode='host',
+                working_dir=self.sandbox_workspace_dir,
+                name=self.container_name,
+                detach=True,
+                volumes={mount_dir: {'bind': self.sandbox_workspace_dir, 'mode': 'rw'}},
+            )
+            print('Container started')
+        except Exception as e:
+            print('Failed to start container')
+            raise e
+
+    @property
+    def sandbox_workspace_dir(self):
+        return config.workspace_mount_path_in_sandbox
+
+    def close(self):
+        containers = self.docker_client.containers.list(all=True)
+        for container in containers:
+            try:
+                if container.name.startswith(self.container_name_prefix):
+                    container.remove(force=True)
+            except docker.errors.NotFound:
+                pass
+        self.docker_client.close()
+    
+    async def on_event(self, event: Event) -> None:
+        print("EventStreamRuntime: on_event triggered")
+        if isinstance(event, Action):
+            observation = await self.run_action(event)
+            print("EventStreamRuntime: observation", observation)
+            # observation._cause = event.id  # type: ignore[attr-defined]
+            source = event.source if event.source else EventSource.AGENT
+            await self.event_stream.add_event(observation, source)
+    
+    async def run_action(self, action: Action) -> Observation:
+        """
+        Run an action and return the resulting observation.
+        If the action is not runnable in any runtime, a NullObservation is returned.
+        If the action is not supported by the current runtime, an ErrorObservation is returned.
+        We will filter some action and execute in runtime. Pass others into od-runtime-client
+        """
+        if not action.runnable:
+            return NullObservation('')
+        action_type = action.action  # type: ignore[attr-defined]
+        if action_type not in ACTION_TYPE_TO_CLASS:
+            return ErrorObservation(f'Action {action_type} does not exist.')
+        if not hasattr(self, action_type):
+            return ErrorObservation(
+                f'Action {action_type} is not supported in the current runtime.'
+            )
+        observation = await getattr(self, action_type)(action)
+        # TODO: fix ID problem
+        observation._parent = action.id  # type: ignore[attr-defined]
```

Comment:
Just a little question here, this is a curious attribute, `_parent`: it's assigned the same way in the server runtime, and it's never used, I think.

There is another attribute, `_cause`, which is relationship between an action and the observation it leads to. It's assigned in the caller of this method in the server runtime. Do you think `_parent` is supposed to have a different meaning than that, or is it possible that we called the same thing with two names at some point during development, and this `_parent` is actually superfluous?

Please note that `_cause` is in actual use on `main`.

Reply:
For this line, I just follow our previous code logic and assign it here. The logic related to other parts, so I didn't dig into too much. I will record a todo here. Let's merge it first and solve those nits in future PR to speed up our progress :).